File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/responses/responseManager.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ActivityTypes } from 'botbuilder';
import { Activity, CardFactory, MessageFactory } from 'botbuilder-core';
import { ActionTypes, Attachment } from 'botframework-schema';
import { readFileSync } from 'fs';
import { getLocale } from 'i18n';
import { join } from 'path';
import { Card } from './card';
import { ICardData } from './cardData';
import { Reply } from './reply';
import { IResponseIdCollection } from './responseIdCollection';
import { ResponseTemplate } from './responseTemplate';

export class ResponseManager {

    private readonly defaultLocaleKey: string = 'default';
    private static readonly simpleTokensRegex: RegExp = /\{(\w+)\}/g;
    private static readonly complexTokensRegex: RegExp = /\{[^{\}]+(?=})\}/g;

    constructor(locales: string[], responseTemplates: IResponseIdCollection[]) {

        this.jsonResponses = new Map();

        responseTemplates.forEach((responseTemplate: IResponseIdCollection) => {
            const resourceName: string = responseTemplate.name;
            const resource: string = responseTemplate.pathToResource || join(__dirname, '..', 'resources');
            this.loadResponses(resourceName, resource);
            locales.forEach((locale: string) => {
                try {
                    this.loadResponses(resourceName, resource, locale);
                } catch {
                    // If satellite assembly doesn't exist, bot will fall back to default.
                }
            });
        });
    }

    public jsonResponses: Map<string, Map<string, ResponseTemplate>>;

    public getResponse(templateId: string, tokens?: Map<string, string>): Partial<Activity> {
        const locale: string = getLocale();
        const template: ResponseTemplate = this.getResponseTemplate(templateId, locale);

        // create the response the data items
        return this.parseResponse(template, tokens);
    }

    public getCardResponse(cards: Card | Card[]): Partial<Activity> {
        const locale: string = getLocale();
        const resourcePath: string = join(__dirname, '..', 'resources', 'cards');

        if (cards instanceof Card) {
            const json: string = this.loadCardJson(cards.name, locale, resourcePath);
            const attachment: Attachment = this.buildCardAttachment(json, cards.data);

            return MessageFactory.attachment(attachment);
        } else {
            const attachments: Attachment[] = [];

            cards.forEach((card: Card) => {
                const json: string = this.loadCardJson(card.name, locale, resourcePath);
                attachments.push(this.buildCardAttachment(json, card.data));
            });

            return MessageFactory.carousel(attachments);
        }
    }

    public getCardResponseWithTemplateId(templateId: string, cards: Card | Card[], tokens?: Map<string, string>): Partial<Activity> {
        const response: Partial<Activity> = this.getResponse(templateId, tokens);
        const locale: string = getLocale();
        const resourcePath: string = join(__dirname, '..', 'resources', 'cards');

        if (cards instanceof Card) {
            const json: string = this.loadCardJson(cards.name, locale, resourcePath);
            const attachment: Attachment = this.buildCardAttachment(json, cards.data);

            return MessageFactory.attachment(attachment, response.text, response.speak, response.inputHint);
        } else {
            const attachments: Attachment[] = [];

            cards.forEach((card: Card) => {
                const json: string = this.loadCardJson(card.name, locale, resourcePath);
                attachments.push(this.buildCardAttachment(json, card.data));
            });

            return MessageFactory.carousel(attachments, response.text, response.speak, response.inputHint);
        }
    }

    public getResponseTemplate(templateId: string, locale?: string): ResponseTemplate {
        let localeKey: string = locale !== undefined ? locale : getLocale();

        // warm up the JsonResponses loading to see if it actually exist.
        // If not, throw with the loading time exception that's actually helpful
        let key: string | undefined = this.getJsonResponseKeyForLocale(templateId, localeKey);

        // if no matching json file found for locale, try parent language
        if (key === undefined) {
            localeKey = localeKey.split('-')[0]
                .toLowerCase();
            key = this.getJsonResponseKeyForLocale(templateId, localeKey);

            // fall back to default
            if (key === undefined) {
                localeKey = this.defaultLocaleKey;
                key = this.getJsonResponseKeyForLocale(templateId, localeKey);
            }
        }

        if (key === undefined) {
            throw new Error();
        }

        // Get the bot response from the .json file
        const responseLocale: Map<string, ResponseTemplate> | undefined = this.jsonResponses.get(localeKey);
        if (!responseLocale || !responseLocale.has(key)) {
            throw new Error(`Unable to find response ${templateId}`);
        }

        const response: ResponseTemplate | undefined = responseLocale.get(key);
        if (response === undefined) {
            throw new Error();
        }

        return response;
    }

    public format(messageTemplate: string, tokens?: Map<string, string>): string {
        let result: string = messageTemplate;

        ResponseManager.complexTokensRegex.lastIndex = 0;
        let match: RegExpExecArray | null = ResponseManager.complexTokensRegex.exec(result);
        while (match) {
            const bindingJson: string = match[0];

            const tokenKey: string = bindingJson
                .replace('{', '')
                .replace('}', '');

            if (tokens && tokens.has(tokenKey)) {
                result = result.replace(bindingJson, tokens.get(tokenKey) || '');
            }

            match = ResponseManager.complexTokensRegex.exec(result);
        }

        return result;
    }

    private loadResponses(resourceName: string, resourcePath: string, locale?: string): void {
        // if locale is not set, add resources under the default key.
        const localeKey: string = (locale !== undefined) ? locale : this.defaultLocaleKey;

        let jsonPath: string = join(resourcePath, `${resourceName}.${localeKey}.json`);

        try {
            require.resolve(jsonPath);
        } catch (errLocale) {
            jsonPath = join(resourcePath, `${resourceName}.json`);

            // Search for the common resource
            try {
                require.resolve(jsonPath);
            } catch (err) {
                throw new Error(`Unable to find '${resourceName}' in '${resourcePath}'`);
            }
        }

        try {
            const content: { [key: string]: Object } = JSON.parse(this.jsonFromFile(jsonPath));

            const localeResponses: Map<string, ResponseTemplate> = this.jsonResponses.get(localeKey) || new Map<string, ResponseTemplate>();

            Object.entries(content)
            .forEach((val: [string, Object]) => {
                const key: string = val[0];
                const value: ITemplate = <ITemplate>val[1];
                const template: ResponseTemplate = Object.assign(new ResponseTemplate(), value);
                localeResponses.set(key, template);
            });

            this.jsonResponses.set(localeKey, localeResponses);
        } catch (err) {
            throw new Error(`Error deserializing ${jsonPath}`);
        }
    }

    private getJsonResponseKeyForLocale(responseId: string, locale: string): string | undefined {
        if (this.jsonResponses.has(locale)) {
            const localeResponses: Map<string, ResponseTemplate> | undefined = this.jsonResponses.get(locale);
            if (localeResponses) {
                return localeResponses.has(responseId) ? responseId : undefined;
            }
        }

        return undefined;
    }

    private parseResponse(template: ResponseTemplate, data?: Map<string, string>): Partial<Activity> {
        const reply: Reply | undefined = template.reply;
        if (!reply) {
            throw new Error('There is no reply in the ResponseTemplate');
        }

        if (reply.text) {
            reply.text = this.format(reply.text, data);
        }

        if (reply.speak) {
            reply.speak = this.format(reply.speak, data);
        }

        const activity: Partial<Activity> = {
            type: ActivityTypes.Message,
            text: reply.text,
            speak: reply.speak,
            inputHint: template.inputHint
        };

        if (template.suggestedActions && template.suggestedActions.length > 0) {
            activity.suggestedActions = {
                actions: [],
                to: []
            };

            template.suggestedActions.forEach((action: string) => {
                if (activity.suggestedActions) {
                    activity.suggestedActions.actions.push({
                        type: ActionTypes.ImBack,
                        title: action,
                        value: action
                    });
                }
            });
        }

        activity.attachments = [];

        return activity;
    }

    private loadCardJson(cardName: string, locale: string, resourcePath: string): string {
        let jsonFile: string = join(resourcePath, `${cardName}.${locale}.json`);

        try {
            require.resolve(jsonFile);
        } catch (errLocale) {
            try {
                jsonFile = join(resourcePath, `${cardName}.json`);
                require.resolve(jsonFile);
            } catch (error) {
                throw new Error(`Could not file Adaptive Card resource ${jsonFile}`);
            }
        }

        return this.jsonFromFile(jsonFile);
    }

    private buildCardAttachment(json: string, data?: ICardData): Attachment {
        let jsonOut: string = json;
        // If cardData was provided
        if (data !== undefined) {
            // add all properties to the list
            const tokens: Map<string, string> = new Map<string, string>();
            // get property names from cardData
            Object.entries(data)
                .forEach((entry: [string, string]) => {
                    if (!tokens.has(entry[0])) {
                        tokens.set(entry[0], entry[1]);
                    }
                });

            // replace tokens in json
            ResponseManager.simpleTokensRegex.lastIndex = 0;
            let match: RegExpExecArray | null = ResponseManager.simpleTokensRegex.exec(jsonOut);
            while (match) {
                if (tokens.has(match[0])) {
                    jsonOut = jsonOut.replace(match[0], tokens.get(match[0]) || '');
                }

                match = ResponseManager.simpleTokensRegex.exec(jsonOut);
            }
        }

        // Deserialize/Serialize logic is needed to prevent JSON exception in prompts
        const cardObj: Object = JSON.parse(jsonOut);

        return CardFactory.adaptiveCard(cardObj);
    }

    private jsonFromFile(filePath: string): string {
        // tslint:disable-next-line:non-literal-fs-path
        return readFileSync(filePath, 'utf8');
    }
}

interface ITemplate {
    replies: Reply[];
    suggestedActions: string[];
    inputHint: string;
}

File: sdk/csharp/libraries/microsoft.bot.solutions/Responses/ResponseManager.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using AdaptiveCards;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Newtonsoft.Json;

namespace Microsoft.Bot.Solutions.Responses
{
    public class ResponseManager
    {
        private const string _defaultLocaleKey = "default";
        private static readonly Regex SimpleTokensRegex = new Regex(@"\{(\w+)\}", RegexOptions.Compiled);
        private static readonly Regex ComplexTokensRegex = new Regex(@"\{[^{\}]+(?=})\}", RegexOptions.Compiled);

        public ResponseManager(string[] locales, params IResponseIdCollection[] responseTemplates)
        {
            JsonResponses = new Dictionary<string, Dictionary<string, ResponseTemplate>>();

            foreach (var responseTemplate in responseTemplates)
            {
                var type = responseTemplate.GetType();
                var resourceName = type.FullName;
                var resourceAssembly = type.Assembly;

                LoadResponses(resourceName, resourceAssembly);

                foreach (var locale in locales)
                {
                    try
                    {
                        var localizedResourceAssembly = resourceAssembly.GetSatelliteAssembly(new CultureInfo(locale));
                        LoadResponses(resourceName, localizedResourceAssembly, locale);
                    }
                    catch
                    {
                        // If satellite assembly doesn't exist, bot will fall back to default.
                    }
                }
            }
        }

        public Dictionary<string, Dictionary<string, ResponseTemplate>> JsonResponses { get; set; }

        /// <summary>
        /// Gets a simple response from template with Text, Speak, InputHint, and SuggestedActions set.
        /// </summary>
        /// <param name="templateId">The name of the response template.</param>
        /// <param name="tokens">StringDictionary of tokens to replace in the response.</param>
        /// <returns>An Activity.</returns>
        public Activity GetResponse(
            string templateId,
            StringDictionary tokens = null)
        {
            var locale = CultureInfo.CurrentUICulture.Name;
            var template = GetResponseTemplate(templateId, locale);

            // create the response the data items
            return ParseResponse(template, tokens);
        }

        /// <summary>
        /// Gets a simple response from template with Text, Speak, InputHint, and SuggestedActions set.
        /// </summary>
        /// <param name="templateId">The name of the response template.</param>
        /// <param name="locale">The locale for the response template.</param>
        /// <param name="tokens">StringDictionary of tokens to replace in the response.</param>
        /// <returns>An Activity.</returns>
        public Activity GetResponse(
            string templateId,
            string locale,
            StringDictionary tokens = null)
        {
            var template = GetResponseTemplate(templateId, locale);

            // create the response the data items
            return ParseResponse(template, tokens);
        }

        /// <summary>
        /// Get a response with an Adaptive Card attachment.
        /// </summary>
        /// <param name="card">The card to add to the response.</param>
        /// <returns>An Activity.</returns>
        public Activity GetCardResponse(Card card)
        {
            var locale = CultureInfo.CurrentUICulture.Name;
            var assembly = Assembly.GetCallingAssembly();
            var json = LoadCardJson(card.Name, locale, assembly);
            var attachment = BuildCardAttachment(json, card.Data);

            return MessageFactory.Attachment(attachment) as Activity;
        }

        /// <summary>
        /// Get a response with a list of Adaptive Card attachments.
        /// </summary>
        /// <param name="cards">The list of Adaptive Cards to add to the response.</param>
        /// <param name="attachmentLayout">Optional AttachmentLayout for resulting activity.</param>
        /// <returns>An Activity.</returns>
        public Activity GetCardResponse(
            IEnumerable<Card> cards,
            string attachmentLayout = AttachmentLayoutTypes.Carousel)
        {
            var locale = CultureInfo.CurrentUICulture.Name;
            var assembly = Assembly.GetCallingAssembly();
            var attachments = new List<Attachment>();

            foreach (var item in cards)
            {
                var json = LoadCardJson(item.Name, locale, assembly);
                attachments.Add(BuildCardAttachment(json, item.Data));
            }

            return MessageFactory.Carousel(attachments) as Activity;
        }

        /// <summary>
        /// Get a response from template with Text, Speak, InputHint, SuggestedActions, and an Adaptive Card attachment.
        /// </summary>
        /// <param name="templateId">The name of the response template.</param>
        /// <param name="card">The card object to add to the response.</param>
        /// <param name="tokens">Optional StringDictionary of tokens to replace in the response.</param>
        /// <returns>An Activity.</returns>
        public Activity GetCardResponse(
            string templateId,
            Card card,
            StringDictionary tokens = null)
        {
            var response = GetResponse(templateId, tokens);
            var locale = CultureInfo.CurrentUICulture.Name;
            var assembly = Assembly.GetCallingAssembly();
            var json = LoadCardJson(card.Name, locale, assembly);
            var attachment = BuildCardAttachment(json, card.Data);

            return MessageFactory.Attachment(attachment, response.Text, response.Speak, response.InputHint) as Activity;
        }

        /// <summary>
        /// Get a response from template with Text, Speak, InputHint, SuggestedActions, and a list of Adaptive Card attachments.
        /// </summary>
        /// <param name="templateId">The name of the response template.</param>
        /// <param name="cards">The collection of Adaptive Cards to add to the response.</param>
        /// <param name="tokens">Optional StringDictionary of tokens to replace in the response.</param>
        /// <param name="attachmentLayout">Optional AttachmentLayout for the resulting activity.</param>
        /// <returns>An Activity.</returns>
        public Activity GetCardResponse(
            string templateId,
            IEnumerable<Card> cards,
            StringDictionary tokens = null,
            string attachmentLayout = AttachmentLayoutTypes.Carousel)
        {
            var response = GetResponse(templateId, tokens);
            var locale = CultureInfo.CurrentUICulture.Name;
            var assembly = Assembly.GetCallingAssembly();
            var attachments = new List<Attachment>();

            foreach (var item in cards)
            {
                var json = LoadCardJson(item.Name, locale, assembly);
                attachments.Add(BuildCardAttachment(json, item.Data));
            }

            return MessageFactory.Carousel(attachments, response.Text, response.Speak, response.InputHint) as Activity;
        }

        /// <summary>
        /// Get a response from template with Text, Speak, InputHint, SuggestedActions, and a Card attachments with list items inside.
        /// </summary>
        /// <param name="templateId">The name of the response template.</param>
        /// <param name="card">The main card container contains list.</param>
        /// <param name="tokens">Optional StringDictionary of tokens to replace in the response.</param>
        /// <param name="containerName">Target container.</param>
        /// <param name="containerItems">Card list which will be injected to target container.</param>
        /// <returns>An Activity.</returns>
        public Activity GetCardResponse(
            string templateId,
            Card card,
            StringDictionary tokens = null,
            string containerName = null,
            IEnumerable<Card> containerItems = null)
        {
            var locale = CultureInfo.CurrentUICulture.Name;
            var assembly = Assembly.GetCallingAssembly();
            var json = LoadCardJson(card.Name, locale, assembly);

            var mainCard = BuildCard(json, card.Data);
            if (!string.IsNullOrEmpty(containerName))
            {
                var itemsContainer = mainCard.Body.Find(item => item.Id == containerName);
                if (itemsContainer is AdaptiveContainer itemsAdaptiveContainer)
                {
                    foreach (var cardItem in containerItems)
                    {
                        var itemJson = LoadCardJson(cardItem.Name, locale, assembly);
                        var itemCard = BuildCard(itemJson, cardItem.Data);
                        foreach (var body in itemCard.Body)
                        {
                            itemsAdaptiveContainer.Items.Add(body);
                        }
                    }
                }
            }

            var cardObj = JsonConvert.DeserializeObject(JsonConvert.SerializeObject(mainCard));
            var attachment = new Attachment(AdaptiveCard.ContentType, content: cardObj);

            if (templateId != null)
            {
                var response = GetResponse(templateId, tokens);
                return MessageFactory.Attachment(attachment, response.Text, response.Speak, response.InputHint) as Activity;
            }

            return MessageFactory.Attachment(attachment, null, null, null) as Activity;
        }

        public ResponseTemplate GetResponseTemplate(string templateId, string locale = null)
        {
            locale = locale ?? CultureInfo.CurrentUICulture.Name;

            // warm up the JsonResponses loading to see if it actually exist. If not, throw with the loading time exception that's actually helpful
            var key = GetJsonResponseKeyForLocale(templateId, locale);

            // if no matching json file found for locale, try parent language
            if (key == null)
            {
                locale = locale.Split('-')[0].ToLower();
                key = GetJsonResponseKeyForLocale(templateId, locale);

                // fall back to default
                if (key == null)
                {
                    locale = _defaultLocaleKey;
                    key = GetJsonResponseKeyForLocale(templateId, locale);
                }
            }

            // Get the bot response from the .json file
            var response = JsonResponses[locale][key ?? throw new KeyNotFoundException($"Unable to find response {templateId}.")];
            return JsonConvert.DeserializeObject<ResponseTemplate>(JsonConvert.SerializeObject(response));
        }

        public string Format(string messageTemplate, StringDictionary tokens)
        {
            var result = messageTemplate;
            var matches = ComplexTokensRegex.Matches(messageTemplate);
            foreach (var match in matches)
            {
                var bindingJson = match.ToString();

                var tokenKey = bindingJson
                  .Replace("{", string.Empty)
                  .Replace("}", string.Empty);

                result = tokens.ContainsKey(tokenKey)
                    ? result.Replace(bindingJson, tokens[tokenKey])
                    : result;
            }

            return result;
        }

        private void LoadResponses(string resourceName, Assembly resourceAssembly, string locale = null)
        {
            var resources = new List<string>();

            // if locale is not set, add resources under the default key.
            var localeKey = _defaultLocaleKey;
            if (locale != null)
            {
                localeKey = new CultureInfo(locale).TwoLetterISOLanguageName;
            }

            var jsonFile = $"{resourceName}.json";

            resources = resourceAssembly
                .GetManifestResourceNames()
                .Where(x => x.Contains(jsonFile))
                .ToList();

            if (resources == null || resources.Count() == 0)
            {
                throw new FileNotFoundException($"Unable to find \"{jsonFile}\" in \"{resourceAssembly.FullName}\" assembly.");
            }

            foreach (var resource in resources)
            {
                try
                {
                    string jsonData;
                    using (var sr = new StreamReader(resourceAssembly.GetManifestResourceStream(resource)))
                    {
                        jsonData = sr.ReadToEnd();
                    }

                    var responses = JsonConvert.DeserializeObject<Dictionary<string, ResponseTemplate>>(jsonData);

                    if (JsonResponses.ContainsKey(localeKey))
                    {
                        var localeResponses = JsonResponses[localeKey] ?? new Dictionary<string, ResponseTemplate>();

                        foreach (var item in responses)
                        {
                            if (!localeResponses.ContainsKey(item.Key))
                            {
                                localeResponses.Add(item.Key, item.Value);
                            }
                        }

                        JsonResponses[localeKey] = localeResponses;
                    }
                    else
                    {
                        JsonResponses.Add(localeKey, responses);
                    }
                }
                catch (JsonReaderException ex)
                {
                    throw new JsonReaderException($"Error deserializing {resource}. {ex.Message}", ex);
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        private string GetJsonResponseKeyForLocale(string responseId, string locale)
        {
            if (JsonResponses.ContainsKey(locale))
            {
                return JsonResponses[locale].ContainsKey(responseId) ? responseId : null;
            }

            return null;
        }

        private Activity ParseResponse(ResponseTemplate template, StringDictionary data)
        {
            var reply = template.Reply;

            if (reply.Text != null)
            {
                reply.Text = this.Format(reply.Text, data);
            }

            if (reply.Speak != null)
            {
                reply.Speak = this.Format(reply.Speak, data);
            }

            var activity = new Activity()
            {
                Type = ActivityTypes.Message,
                Text = reply.Text,
                Speak = reply.Speak,
                InputHint = template.InputHint,
            };

            if (template.SuggestedActions != null && template.SuggestedActions.Count() > 0)
            {
                activity.SuggestedActions = new SuggestedActions
                {
                    Actions = new List<CardAction>(),
                };

                foreach (var action in template.SuggestedActions)
                {
                    activity.SuggestedActions.Actions.Add(new CardAction(type: ActionTypes.ImBack, title: action, value: action));
                }
            }

            activity.Attachments = new List<Attachment>();

            return activity;
        }

        private string LoadCardJson(string cardName, string locale, Assembly assembly)
        {
            // get card json for locale
            var culture = new CultureInfo(locale);
            var langCode = culture.TwoLetterISOLanguageName;
            var jsonFile = $"{cardName}.json";
            var json = string.Empty;
            var resource = string.Empty;

            try
            {
                // try to get the localized json
                resource = assembly
                    .GetSatelliteAssembly(culture)
                    .GetManifestResourceNames()
                    .Where(x => x.Contains(jsonFile))
                    .First();
            }
            catch (FileNotFoundException)
            {
                // If the localized file is missing, try falling back to the default language.
                resource = assembly
                    .GetManifestResourceNames()
                    .Where(x => x.Contains(jsonFile))
                    .First();
            }
            catch (Exception)
            {
                throw new Exception($"Could not file Adaptive Card resource {jsonFile}");
            }

            using (var sr = new StreamReader(assembly.GetManifestResourceStream(resource)))
            {
                json = sr.ReadToEnd();
            }

            return json;
        }

        private Attachment BuildCardAttachment(string json, ICardData data = null)
        {
            var card = BuildCard(json, data);
            var cardObj = JsonConvert.DeserializeObject(JsonConvert.SerializeObject(card));
            return new Attachment(AdaptiveCard.ContentType, content: cardObj);
        }

        private AdaptiveCard BuildCard(string json, ICardData data = null)
        {
            // If cardData was provided
            if (data != null)
            {
                // get property names from cardData
                var properties = data.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);

                // add all properties to the list
                var tokens = new StringDictionary();
                foreach (var property in properties)
                {
                    if (!tokens.ContainsKey(property.Name))
                    {
                        var escapedTokenStr = property.GetValue(data)?.ToString()?.Replace("\\", "\\\\");
                        escapedTokenStr = escapedTokenStr?.Replace("\"", "\\\"");
                        escapedTokenStr = escapedTokenStr?.Replace("\'", "\\\'");
                        tokens.Add(property.Name, escapedTokenStr);
                    }
                }

                // replace tokens in json
                if (tokens != null)
                {
                    json = SimpleTokensRegex.Replace(json, match => tokens[match.Groups[1].Value]);
                }
            }

            // Deserialize/Serialize logic is needed to prevent JSON exception in prompts
            var card = AdaptiveCard.FromJson(json).Card;
            return card;
        }
    }
}